examples/dcp/stateful_example.py

```diff
@@ -0,0 +1,138 @@
+import torch.multiprocessing as mp
+import torch.nn as nn
+from torch.distributed.checkpoint.state_dict import (
+    _patch_model_state_dict,
+    _patch_optimizer_state_dict,
+)
+from torch.distributed.device_mesh import init_device_mesh
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+from s3torchconnector.dcp import S3StorageWriter, S3StorageReader
+
+
+class Model(torch.nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        torch.manual_seed(0)
+        self.net1 = nn.Sequential(nn.Linear(8, 16), nn.ReLU())
+        self.net2 = nn.Sequential(nn.Linear(16, 32), nn.ReLU())
+        self.net3 = nn.Linear(32, 64)
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(64, 8))
+
+    def forward(self, x):
+        return self.net4(self.net3(self.net2(self.net1(x))))
+
+    def get_input(self):
+        return torch.rand(8, 8, device="cuda")
+
+
+def _make_stateful(model, optim):
+    _patch_model_state_dict(model)
+    _patch_optimizer_state_dict(model, optimizers=optim)
+
+
+def _train(model, optim, train_steps=1):
+    torch.manual_seed(0)
+    loss = None
+    for _ in range(train_steps):
+        loss = model(model.get_input()).sum()
+        loss.backward()
+        optim.step()
+        optim.zero_grad()
+
+    return loss
+
+
+def _init_model(device, world_size):
+    device_mesh = init_device_mesh(device, (world_size,))
+    model = Model().cuda()
+    model = FSDP(
+        model,
+        device_mesh=device_mesh,
+        use_orig_params=True,
+    )
+    optim = torch.optim.Adam(model.parameters(), lr=0.1)
+    _make_stateful(model, optim)
+
+    return model, optim
+
+def _compare_models(model1, model2, rank, rtol=1e-5, atol=1e-8):
+    model1.eval()
+    model2.eval()
+
+    with FSDP.summon_full_params(model1), FSDP.summon_full_params(model2):
+        for (name1, param1), (name2, param2) in zip(model1.named_parameters(), model2.named_parameters()):
+            if name1 != name2:
+                print(f"Parameter names don't match: {name1} vs {name2}. Rank:{rank}")
+                return False
+
+            if not torch.allclose(param1, param2, rtol=rtol, atol=atol):
+                print(f"Parameters don't match for {name1}. Rank:{rank}")
+                print(f"Max difference: {(param1 - param2).abs().max().item()}. Rank:{rank}")
+                return False
+
+    print("All parameters match within the specified tolerance. Rank:{rank}")
+    return True
+
+def run(rank, world_size, device="cuda"):
+    # Set up world pg
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+
+    dist.init_process_group("cpu:gloo,cuda:nccl", rank=rank, world_size=world_size)
+    torch.cuda.set_device(rank)
+
+    print(f"Train initial model on rank:{rank}")
+    model, optim = _init_model(device, world_size)
+    _train(model, optim, train_steps=2)
+
+    region = "us-east-1"
+    s3_uri = f"s3://PLEASE-PROVIDE-ACTUAL-BUCKET-NAME/"
```

Comment:
nit: could be uppercase variables at the top of the script, so user would know from the top they need to fill that up; also, unnecessary `f` at the front

Reply:
Agree with removing `f`. Regarding location, I think it is better to keep it close to first usage of `S3StorageWriter`, that reader won't be needed to jump around file for looking what these variables mean.

Reply:
I changed it to be an method's parameters to be on the same line with lighting samples